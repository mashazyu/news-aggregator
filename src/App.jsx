import { useQuery } from "@tanstack/react-query";

import Article from "./components/Article";
import { getArticles } from "./api/articles";

function App() {
  const articlesQuery = useQuery({
    queryKey: ["articles"],
    queryFn: getArticles,
  });

  const { data, isLoading, isError, error } = articlesQuery;
  const { articles } = data;

  if (isLoading) return <h1>loading...</h1>;
  if (isError) return <h1>{JSON.stringify(error)}</h1>;

  return (
    <main className="container mx-auto">
      <div className="grid grid-cols-3 gap-8">
        {articles.map((article) => (
          <Article article={article} key={article.title} />
        ))}
      </div>
    </main>
  );
}

export default App;
