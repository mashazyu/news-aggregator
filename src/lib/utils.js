import { clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs) {
  return twMerge(clsx(inputs));
}

export const isCreatorAvailable = (creator) => {
  return creator?.length > 0 && creator[0]?.length > 0;
};

export const getCreator = (creator) => {
  return isCreatorAvailable(creator) ? creator[0] : "";
};

export const hoursAgo = (dateString) => {
  if (!dateString || isNaN(Date.parse(dateString))) return "";

  const eventDate = new Date(dateString);
  const now = new Date();
  const diffInMs = now - eventDate;
  const hours = Math.floor(diffInMs / (1000 * 60 * 60));

  return hours > 0
    ? `${hours} hours ago `
    : `${diffInMs / (1000 * 60)} minutes ago `;
};

export const limitChars = (text, limit = 100) => {
  if (!text) return "";

  return text.length > limit ? text.substring(0, limit) + "..." : text;
};
