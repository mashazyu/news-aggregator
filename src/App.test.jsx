import { describe, it, expect, vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import axios from "axios";

import { wrapper } from "../tests/utils";
import { articleMock } from "../tests/mocks";
import App from "./App";

vi.spyOn(axios, "get").mockImplementation(() =>
  Promise.resolve({
    data: {
      articles: [articleMock],
      status: "ok",
      totalResults: 1,
    },
  })
);

describe("App", () => {
  it("renders articles", async () => {
    const { asFragment } = render(<App />, { wrapper });

    await waitFor(() => expect(screen.getAllByText("Title")).toHaveLength(1));

    expect(asFragment()).toMatchSnapshot();
  });
});
