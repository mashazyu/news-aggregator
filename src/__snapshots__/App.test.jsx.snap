// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders articles with correct styling 1`] = `
<DocumentFragment>
  <main
    class="container mx-auto py-8"
  >
    <input
      class="flex h-10 w-full rounded-md border border-slate-200 bg-white px-3 py-2 text-sm ring-offset-white file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:bg-slate-950 dark:ring-offset-slate-950 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300"
      placeholder="Filter your news"
      value=""
    />
    <div
      class="flex py-8"
    >
      <div
        class="flex items-center gap-1 flex-wrap justify-start"
        dir="ltr"
        role="group"
        style="outline: none;"
        tabindex="0"
      >
        <button
          aria-checked="true"
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-slate-100 data-[state=on]:text-slate-900 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=on]:bg-slate-800 dark:data-[state=on]:text-slate-50 border border-slate-200 bg-transparent hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:hover:bg-slate-800 dark:hover:text-slate-50 h-10 px-3"
          data-radix-collection-item=""
          data-state="on"
          role="radio"
          tabindex="-1"
          type="button"
        >
          BUSINESS
        </button>
        <button
          aria-checked="false"
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-slate-100 data-[state=on]:text-slate-900 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=on]:bg-slate-800 dark:data-[state=on]:text-slate-50 border border-slate-200 bg-transparent hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:hover:bg-slate-800 dark:hover:text-slate-50 h-10 px-3"
          data-radix-collection-item=""
          data-state="off"
          role="radio"
          tabindex="-1"
          type="button"
        >
          ENTERTAINMENT
        </button>
        <button
          aria-checked="false"
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-slate-100 data-[state=on]:text-slate-900 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=on]:bg-slate-800 dark:data-[state=on]:text-slate-50 border border-slate-200 bg-transparent hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:hover:bg-slate-800 dark:hover:text-slate-50 h-10 px-3"
          data-radix-collection-item=""
          data-state="off"
          role="radio"
          tabindex="-1"
          type="button"
        >
          GENERAL
        </button>
        <button
          aria-checked="false"
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-slate-100 data-[state=on]:text-slate-900 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=on]:bg-slate-800 dark:data-[state=on]:text-slate-50 border border-slate-200 bg-transparent hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:hover:bg-slate-800 dark:hover:text-slate-50 h-10 px-3"
          data-radix-collection-item=""
          data-state="off"
          role="radio"
          tabindex="-1"
          type="button"
        >
          HEALTH
        </button>
        <button
          aria-checked="false"
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-slate-100 data-[state=on]:text-slate-900 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=on]:bg-slate-800 dark:data-[state=on]:text-slate-50 border border-slate-200 bg-transparent hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:hover:bg-slate-800 dark:hover:text-slate-50 h-10 px-3"
          data-radix-collection-item=""
          data-state="off"
          role="radio"
          tabindex="-1"
          type="button"
        >
          SCIENCE
        </button>
        <button
          aria-checked="false"
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-slate-100 data-[state=on]:text-slate-900 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=on]:bg-slate-800 dark:data-[state=on]:text-slate-50 border border-slate-200 bg-transparent hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:hover:bg-slate-800 dark:hover:text-slate-50 h-10 px-3"
          data-radix-collection-item=""
          data-state="off"
          role="radio"
          tabindex="-1"
          type="button"
        >
          SPORTS
        </button>
        <button
          aria-checked="false"
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-slate-100 data-[state=on]:text-slate-900 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 dark:data-[state=on]:bg-slate-800 dark:data-[state=on]:text-slate-50 border border-slate-200 bg-transparent hover:bg-slate-100 hover:text-slate-900 dark:border-slate-800 dark:hover:bg-slate-800 dark:hover:text-slate-50 h-10 px-3"
          data-radix-collection-item=""
          data-state="off"
          role="radio"
          tabindex="-1"
          type="button"
        >
          TECHNOLOGY
        </button>
      </div>
    </div>
    <div
      class="shrink-0 bg-slate-200 dark:bg-slate-800 h-[1px] w-full"
      data-orientation="horizontal"
      role="none"
    />
    <div
      class="grid grid-cols-1 sm:grid-cols-2 gap-8 py-8"
    >
      <div
        class="rounded-lg border border-slate-200 bg-white text-slate-950 shadow-sm dark:border-slate-800 dark:bg-slate-950 dark:text-slate-50 flex flex-col justify-between"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="text-2xl font-semibold leading-none tracking-tight"
          >
            Title 1
          </h3>
          <p
            class="text-sm text-slate-500 dark:text-slate-400"
          >
            by Author
          </p>
        </div>
        <div
          class="p-6 pt-0"
        >
          <p>
            Description
          </p>
        </div>
        <div
          class="flex items-center p-6 pt-0"
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-slate-100 text-slate-900 hover:bg-slate-100/80 dark:bg-slate-800 dark:text-slate-50 dark:hover:bg-slate-800/80 h-10 px-4 py-2"
            href="https://mock.url"
            target="_blank"
          >
            Learn more
          </a>
        </div>
      </div>
      <div
        class="rounded-lg border border-slate-200 bg-white text-slate-950 shadow-sm dark:border-slate-800 dark:bg-slate-950 dark:text-slate-50 flex flex-col justify-between"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="text-2xl font-semibold leading-none tracking-tight"
          >
            Title 2
          </h3>
          <p
            class="text-sm text-slate-500 dark:text-slate-400"
          >
            by Author
          </p>
        </div>
        <div
          class="p-6 pt-0"
        >
          <p>
            Description
          </p>
        </div>
        <div
          class="flex items-center p-6 pt-0"
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-slate-100 text-slate-900 hover:bg-slate-100/80 dark:bg-slate-800 dark:text-slate-50 dark:hover:bg-slate-800/80 h-10 px-4 py-2"
            href="https://mock.url"
            target="_blank"
          >
            Learn more
          </a>
        </div>
      </div>
      <div
        class="rounded-lg border border-slate-200 bg-white text-slate-950 shadow-sm dark:border-slate-800 dark:bg-slate-950 dark:text-slate-50 flex flex-col justify-between"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="text-2xl font-semibold leading-none tracking-tight"
          >
            Title 3
          </h3>
          <p
            class="text-sm text-slate-500 dark:text-slate-400"
          >
            by Author
          </p>
        </div>
        <div
          class="p-6 pt-0"
        >
          <p>
            Description
          </p>
        </div>
        <div
          class="flex items-center p-6 pt-0"
        >
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-slate-100 text-slate-900 hover:bg-slate-100/80 dark:bg-slate-800 dark:text-slate-50 dark:hover:bg-slate-800/80 h-10 px-4 py-2"
            href="https://mock.url"
            target="_blank"
          >
            Learn more
          </a>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
